lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,178 @@
+---------------------------------------------------------------------------
+--- Querying current GTK+ 3 theme via GtkStyleContext.
+--
+-- @author Yauheni Kirylau &lt;yawghen@gmail.com&gt;
+-- @copyright 2016-2017 Yauheni Kirylau
+-- @module beautiful.gtk
+---------------------------------------------------------------------------
+local gears_debug = require("gears.debug")
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+local gtk = {
+  cached_theme_variables = nil
+}
+
+
+local function convert_gtk_color_to_hex(gtk_color)
+  local rgb = gtk_color:to_string()
+  local m = rgb:gmatch("[0-9]+")
+  return m and (
+    "#" .. string.format("%2.2x", m()) .. string.format("%2.2x", m()) .. string.format("%2.2x", m())
+  )
+end
+
+
+local function lookup_gtk_color_to_hex(_style_context, color_name)
+  local color = _style_context:lookup_color(color_name)
+  if not color then
+      return nil
+  end
+  local result = convert_gtk_color_to_hex(color)
+  --color:free()
+  return result
+end
+
+local function get_gtk_color_property_to_hex(_style_context, property_name, state)
+    local property = _style_context:get_property(property_name, state)
+    if not property then
+      return nil
+    end
+    local result = convert_gtk_color_to_hex(property.value)
+    property:unset()
+    return result
```

Comment:
First: I think that `:unset` is unnecessary and pointless.
Second: Wait, what is the difference between this function and the function before? They seem to do the same thing to me, except that this one seems much more scare when looking at the API docs for `GValue`.
Are these two difference ways really necessary? That sounds a lot like this code tries to do something that should not be done.

Reply:
because properties are the real properties of object, while stylecontext colors are just variables which are defined inside stylecontext.

however if theme is using pixmaps or gradient the properties will be not very useful (especially in case of pixmaps) so here come stylecontext

Reply:
also in the docs it's recommended to unset it[1]:

```
When value is no longer needed, g_value_unset() must be called to free any allocated memory.
```

[1] https://developer.gnome.org/gtk3/stable/GtkStyleContext.html#gtk-style-context-get-property

Reply:
Well, yeah, in C code. Since Lua is supposed to be garbage collected, I would expect that lgi does the call to `:unset()` for us. From a quick `git grep unset` through lgi, I would guess that line 54 in `lgi/override/GObject-Value.lua` sets this up: `Value._uninit = core.record.value_unset`
Since colors are not huge objects whose size is invisible to the GC (in contrast to some of the cairo surface that causes us lots of problems and made us add calls to `:finish()` in various places), I really think these `:unset()`-calls can be dropped.